### PR TITLE
feat(internal): Added RecycledQueue class

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

This PR introduces `RecycledQueue` class, which is similar to a regular FIFO `Queue` except that it supports recycling of objects instead of having to create/delete them all the time.

This class will be used to build more efficient component lifecycle queues.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.

## Breaking Change

- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

File: packages/flame/lib/src/utils/recycled_queue.dart [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+  /// The backing container.
+  ///
+  /// Some of the items in this list are considered "active" while others are
+  /// "disposed". The [_startIndex] and [_endIndex] describe the range of items
+  /// in the list which are "active".
+  ///
+  /// Two data layouts are possible: the normal one, when [_endIndex] is larger
+  /// than or equal to the [_startIndex]:
+  ///   [----##############--]
```
Comment (spydon) [inline packages/flame/lib/src/utils/recycled_queue.dart:43]: Do these not have to be in back ticks for the doc parser not to regard them as references to fields?

**Review by spydon — APPROVED** [review]
Lgtm! Not too fond of the utils directory becoming a place with a wide mix of things in it though, maybe the two existing files in there could be put in some `math` directory, and this one be placed somewhere else or having the utils directory renamed?

File: packages/flame/lib/src/utils/recycled_queue.dart (hunk 1 above)
Comment (st-pasha) [reply packages/flame/lib/src/utils/recycled_queue.dart]: I've added the backticks.

**Comment by st-pasha** [comment]
> and this one be placed somewhere else or having the utils directory renamed?

To me, "utils" sounds a bit like "miscellaneous" (but easier to spell :).
So, where would you suggest the `RecycledQueue` class be located?

**Comment by spydon** [comment]
> To me, "utils" sounds a bit like "miscellaneous" (but easier to spell :). So, where would you suggest the `RecycledQueue` class be located?

Indeed, that's what I'd like to avoid, to not have a place where non-related files live.
Maybe a new directory in `components` called `lifecycle`? And then maybe the `LifecycleManager` could be moved there too?

**Comment by st-pasha** [comment]
I created the directory `components/core` then, seems like it would have a larger scope than mere `lifecycle`.

**Comment by spydon** [comment]
> I created the directory `components/core` then, seems like it would have a larger scope than mere `lifecycle`.

Very nice!

**Review by erickzanardo — APPROVED** [review]

**Pull request merged**